iree/compiler/Utils/CustomKernelsTargetInfo.h

```diff
@@ -0,0 +1,90 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_COMPILER_UTILS_CUSTOMKERNELTARGETINFO_H_
+#define IREE_COMPILER_UTILS_CUSTOMKERNELTARGETINFO_H_
+
+#include <stdint.h>
+
+#include <cassert>
+
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Support/LogicalResult.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+// Enumerates target ISAs that we care about.
+enum class CustomKernelTargetArch { None, Aarch64 };
+
+// Enumerates arch-specific target features that we care about.
+// We explicitly want to stick to the default enumeration values (0, 1, 2, ...,
+// no greater than 63) because this is going to be indexing a uint64 bitfield.
+// Intentionally not reusing bits across architectures to be able to catch
+// most bugs. 64 is enough across all target architectures for now.
+enum class CustomKernelTargetFeature {
+  // Indicates a preference for intrinsics over inline asm. Unlike other bits,
+  // this is generic, not tied to a particular architecture or CPU feature, and
+  // it has to be passed through some custom boolean flag or option, not as
+  // part of the target CPU features.
+  Intrinsics,
+  // Aarch64 features.
+  Aarch64Dotprod,
+};
```

Comment:
Great question! I actually started out this way, as that's what I had [done](https://github.com/google/ruy/blob/master/ruy/path.h#L55) in ruy.

One immediate cost of this approach (with C++ scoped enums, so unlike your above link which is C) is that since scoped enums are strongly typed and do not have bitwise operators readily defined on them, one has to define them manually, which adds a little bit of unwanted details to this file, like [this](https://github.com/google/ruy/blob/master/ruy/path.h#L92-L109). Not a big deal though.

In the end what decided me to go the opposite route here is that this is being used in a very constrained way. The user never needs to know that it's a bitfield, that's an internal detail. From the user's perspective, we are implementing something akin to a set. Since the user doesn't need to know, we might as well conceal that detail. This brings the safety of enforcing in the type system that one is dealing with one feature at a time, without having to assert that values are powers of two. Ruy was different in that the users needed the OR combinations.